Panic in fake client when updating status with Unstructured resource

There is a regression in 0.16.2 in the fake client when attempting to update a resource's status with an unstructured object.

In this case, the GVK has a type registered in the scheme, but the client is invoked using an unstructured object. The initial object was also defined as unstructured and [converted to the typed resource inside the tracker](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.16.2/pkg/client/fake/client.go#L292).

The panic is triggered at
https://github.com/kubernetes-sigs/controller-runtime/blob/v0.16.2/pkg/client/fake/client.go#L407

- `obj` is the unstructured value passed to `client.Status().Update()`
- `oldObject` is a typed value managed by the tracker

The two types are incompatible and cannot be reflectively assigned, at least not naively.

```
panic: reflect.Set: value of type v1.MyResource is not assignable to type unstructured.Unstructured [recovered]
	panic: reflect.Set: value of type v1.MyResource is not assignable to type unstructured.Unstructured

goroutine 12 [running]:
testing.tRunner.func1.2({0x105d91e20, 0x140004cc010})
	/opt/homebrew/Cellar/go/1.21.1/libexec/src/testing/testing.go:1545 +0x370
testing.tRunner.func1()
	/opt/homebrew/Cellar/go/1.21.1/libexec/src/testing/testing.go:1548 +0x528
panic({0x105d91e20?, 0x140004cc010?})
	/opt/homebrew/Cellar/go/1.21.1/libexec/src/runtime/panic.go:920 +0x254
reflect.Value.assignTo({0x105f00420, 0x14000100900, 0x199}, {0x105b0d7e6, 0xb}, 0x105e6d040, 0x0)
	/opt/homebrew/Cellar/go/1.21.1/libexec/src/reflect/value.go:3307 +0x340
reflect.Value.Set({0x105e6d040, 0x14000683588, 0x199}, {0x105f00420, 0x14000100900, 0x199})
	/opt/homebrew/Cellar/go/1.21.1/libexec/src/reflect/value.go:2260 +0xa8
sigs.k8s.io/controller-runtime/pkg/client/fake.versionedTracker.update({{0x1060581a0, 0x140000a0370}, 0x140002885b0, 0x1400007fe30}, {{0x1400069dde0, 0x1a}, {0x1400069ddfb, 0x2}, {0x140001ae220, 0xd}}, ...)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.2/pkg/client/fake/client.go:407 +0x938
sigs.k8s.io/controller-runtime/pkg/client/fake.(*fakeClient).update(0x140000a03c0, {0x10605dd18, 0x14000683588}, 0x1, {0x14000356dc0, 0x1, 0x1})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.2/pkg/client/fake/client.go:789 +0x46c
sigs.k8s.io/controller-runtime/pkg/client/fake.(*fakeSubResourceClient).Update(0x14000516f18, {0x106054c50, 0x14000024c30}, {0x10605dd18, 0x14000683588}, {0x0, 0x0, 0x0})
```

--- a/pkg/client/fake/client.go
+++ b/pkg/client/fake/client.go
@@ -404,7 +404,9 @@ func (t versionedTracker) update(gvr schema.GroupVersionResource, obj runtime.Ob
 				return fmt.Errorf("failed to copy non-status field for object with status subresouce: %w", err)
 			}
 			passedRV := accessor.GetResourceVersion()
-			reflect.ValueOf(obj).Elem().Set(reflect.ValueOf(oldObject.DeepCopyObject()).Elem())
+			if err := copyFrom(oldObject, obj); err != nil {
+				return fmt.Errorf("failed to restore non-status fields: %w", err)
+			}
 			accessor.SetResourceVersion(passedRV)
 		} else { // copy status from original object
 			if err := copyStatusFrom(oldObject, obj); err != nil {
@@ -966,6 +968,19 @@ func copyStatusFrom(old, new runtime.Object) error {
 	newMapStringAny["status"] = oldMapStringAny["status"]
 
 	if err := fromMapStringAny(newMapStringAny, new); err != nil {
+		return fmt.Errorf("failed to convert back from map[string]any: %w", err)
+	}
+
+	return nil
+}
+
+// copyFrom copies from old into new
+func copyFrom(old, new runtime.Object) error {
+	oldMapStringAny, err := toMapStringAny(old)
+	if err != nil {
+		return fmt.Errorf("failed to convert old to *unstructured.Unstructured: %w", err)
+	}
+	if err := fromMapStringAny(oldMapStringAny, new); err != nil {
 		return fmt.Errorf("failed to convert back from map[string]any: %w", err)
 	}
 

--- a/pkg/client/fake/client_test.go
+++ b/pkg/client/fake/client_test.go
@@ -1677,6 +1677,52 @@ var _ = Describe("Fake client", func() {
 		Expect(obj.Object["spec"]).To(BeEquivalentTo("original"))
 	})
 
+	It("should not change the status of known unstructured objects that have a status subresource on update", func() {
+		obj := &unstructured.Unstructured{}
+		obj.SetAPIVersion("v1")
+		obj.SetKind("Pod")
+		obj.SetName("pod")
+		err := unstructured.SetNestedField(obj.Object, "Always", "spec", "restartPolicy")
+		Expect(err).NotTo(HaveOccurred())
+		err = unstructured.SetNestedField(obj.Object, "Pending", "status", "phase")
+		Expect(err).NotTo(HaveOccurred())
+		cl := NewClientBuilder().WithStatusSubresource(obj).WithObjects(obj).Build()
+
+		err = unstructured.SetNestedField(obj.Object, "Never", "spec", "restartPolicy")
+		Expect(err).NotTo(HaveOccurred())
+		err = unstructured.SetNestedField(obj.Object, "Running", "status", "phase")
+		Expect(err).NotTo(HaveOccurred())
+
+		Expect(cl.Update(context.Background(), obj)).To(Succeed())
+		Expect(cl.Get(context.Background(), client.ObjectKeyFromObject(obj), obj)).To(Succeed())
+
+		Expect(obj.Object["spec"].(map[string]any)["restartPolicy"]).To(Equal("Never"))
+		Expect(obj.Object["status"].(map[string]any)["phase"]).To(Equal("Pending"))
+	})
+
+	It("should not change non-status field of known unstructured objects that have a status subresource on status update", func() {
+		obj := &unstructured.Unstructured{}
+		obj.SetAPIVersion("v1")
+		obj.SetKind("Pod")
+		obj.SetName("pod")
+		err := unstructured.SetNestedField(obj.Object, "Always", "spec", "restartPolicy")
+		Expect(err).NotTo(HaveOccurred())
+		err = unstructured.SetNestedField(obj.Object, "Pending", "status", "phase")
+		Expect(err).NotTo(HaveOccurred())
+		cl := NewClientBuilder().WithStatusSubresource(obj).WithObjects(obj).Build()
+
+		err = unstructured.SetNestedField(obj.Object, "Never", "spec", "restartPolicy")
+		Expect(err).NotTo(HaveOccurred())
+		err = unstructured.SetNestedField(obj.Object, "Running", "status", "phase")
+		Expect(err).NotTo(HaveOccurred())
+
+		Expect(cl.Status().Update(context.Background(), obj)).To(Succeed())
+		Expect(cl.Get(context.Background(), client.ObjectKeyFromObject(obj), obj)).To(Succeed())
+
+		Expect(obj.Object["spec"].(map[string]any)["restartPolicy"]).To(Equal("Always"))
+		Expect(obj.Object["status"].(map[string]any)["phase"]).To(Equal("Running"))
+	})
+
 	It("should not change the status of unstructured objects that are configured to have a status subresource on patch", func() {
 		obj := &unstructured.Unstructured{}
 		obj.SetAPIVersion("foo/v1")

--- a/pkg/client/fake/client_test.go
+++ b/pkg/client/fake/client_test.go
@@ -1678,49 +1678,81 @@ var _ = Describe("Fake client", func() {
 	})
 
 	It("should not change the status of known unstructured objects that have a status subresource on update", func() {
-		obj := &unstructured.Unstructured{}
-		obj.SetAPIVersion("v1")
-		obj.SetKind("Pod")
-		obj.SetName("pod")
-		err := unstructured.SetNestedField(obj.Object, "Always", "spec", "restartPolicy")
-		Expect(err).NotTo(HaveOccurred())
-		err = unstructured.SetNestedField(obj.Object, "Pending", "status", "phase")
-		Expect(err).NotTo(HaveOccurred())
+		obj := &corev1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "pod",
+			},
+			Spec: corev1.PodSpec{
+				RestartPolicy: corev1.RestartPolicyAlways,
+			},
+			Status: corev1.PodStatus{
+				Phase: corev1.PodPending,
+			},
+		}
 		cl := NewClientBuilder().WithStatusSubresource(obj).WithObjects(obj).Build()
 
-		err = unstructured.SetNestedField(obj.Object, "Never", "spec", "restartPolicy")
-		Expect(err).NotTo(HaveOccurred())
-		err = unstructured.SetNestedField(obj.Object, "Running", "status", "phase")
+		// update using unstructured
+		u := &unstructured.Unstructured{}
+		u.SetAPIVersion("v1")
+		u.SetKind("Pod")
+		u.SetName(obj.Name)
+		err := cl.Get(context.Background(), client.ObjectKeyFromObject(u), u)
 		Expect(err).NotTo(HaveOccurred())
 
-		Expect(cl.Update(context.Background(), obj)).To(Succeed())
-		Expect(cl.Get(context.Background(), client.ObjectKeyFromObject(obj), obj)).To(Succeed())
+		err = unstructured.SetNestedField(u.Object, string(corev1.RestartPolicyNever), "spec", "restartPolicy")
+		Expect(err).NotTo(HaveOccurred())
+		err = unstructured.SetNestedField(u.Object, string(corev1.PodRunning), "status", "phase")
+		Expect(err).NotTo(HaveOccurred())
 
-		Expect(obj.Object["spec"].(map[string]any)["restartPolicy"]).To(Equal("Never"))
-		Expect(obj.Object["status"].(map[string]any)["phase"]).To(Equal("Pending"))
+		Expect(cl.Update(context.Background(), u)).To(Succeed())
+
+		actual := &corev1.Pod{}
+		Expect(cl.Get(context.Background(), client.ObjectKeyFromObject(obj), actual)).To(Succeed())
+		obj.APIVersion = u.GetAPIVersion()
+		obj.Kind = u.GetKind()
+		obj.ResourceVersion = actual.ResourceVersion
+		// only the spec mutation should persist
+		obj.Spec.RestartPolicy = corev1.RestartPolicyNever
+		Expect(cmp.Diff(obj, actual)).To(BeEmpty())
 	})
 
 	It("should not change non-status field of known unstructured objects that have a status subresource on status update", func() {
-		obj := &unstructured.Unstructured{}
-		obj.SetAPIVersion("v1")
-		obj.SetKind("Pod")
-		obj.SetName("pod")
-		err := unstructured.SetNestedField(obj.Object, "Always", "spec", "restartPolicy")
-		Expect(err).NotTo(HaveOccurred())
-		err = unstructured.SetNestedField(obj.Object, "Pending", "status", "phase")
-		Expect(err).NotTo(HaveOccurred())
+		obj := &corev1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "pod",
+			},
+			Spec: corev1.PodSpec{
+				RestartPolicy: corev1.RestartPolicyAlways,
+			},
+			Status: corev1.PodStatus{
+				Phase: corev1.PodPending,
+			},
+		}
 		cl := NewClientBuilder().WithStatusSubresource(obj).WithObjects(obj).Build()
 
-		err = unstructured.SetNestedField(obj.Object, "Never", "spec", "restartPolicy")
-		Expect(err).NotTo(HaveOccurred())
-		err = unstructured.SetNestedField(obj.Object, "Running", "status", "phase")
+		// status update using unstructured
+		u := &unstructured.Unstructured{}
+		u.SetAPIVersion("v1")
+		u.SetKind("Pod")
+		u.SetName(obj.Name)
+		err := cl.Get(context.Background(), client.ObjectKeyFromObject(u), u)
 		Expect(err).NotTo(HaveOccurred())
 
-		Expect(cl.Status().Update(context.Background(), obj)).To(Succeed())
-		Expect(cl.Get(context.Background(), client.ObjectKeyFromObject(obj), obj)).To(Succeed())
+		err = unstructured.SetNestedField(u.Object, string(corev1.RestartPolicyNever), "spec", "restartPolicy")
+		Expect(err).NotTo(HaveOccurred())
+		err = unstructured.SetNestedField(u.Object, string(corev1.PodRunning), "status", "phase")
+		Expect(err).NotTo(HaveOccurred())
 
-		Expect(obj.Object["spec"].(map[string]any)["restartPolicy"]).To(Equal("Always"))
-		Expect(obj.Object["status"].(map[string]any)["phase"]).To(Equal("Running"))
+		Expect(cl.Status().Update(context.Background(), u)).To(Succeed())
+
+		actual := &corev1.Pod{}
+		Expect(cl.Get(context.Background(), client.ObjectKeyFromObject(obj), actual)).To(Succeed())
+		obj.APIVersion = "v1"
+		obj.Kind = "Pod"
+		obj.ResourceVersion = actual.ResourceVersion
+		// only the status mutation should persist
+		obj.Status.Phase = corev1.PodRunning
+		Expect(cmp.Diff(obj, actual)).To(BeEmpty())
 	})
 
 	It("should not change the status of unstructured objects that are configured to have a status subresource on patch", func() {

